File: src/main/java/io/rong/methods/profile/Friend.java
package io.rong.methods.profile;

import io.rong.RongCloud;
import io.rong.methods.BaseMethod;
import io.rong.models.CheckMethod;
import io.rong.models.profile.FriendModel;
import io.rong.models.profile.FriendProfileModel;
import io.rong.models.profile.PagingGetFriendsModel;
import io.rong.models.response.CheckFriendsResult;
import io.rong.models.response.GetPermissionResult;
import io.rong.models.response.QueryFriendsResult;
import io.rong.models.response.ResponseResult;
import org.apache.commons.lang3.StringUtils;

/**
 * Group Information Hosting Service
 */
public class Friend extends BaseMethod {

    private static final String API_JSON_PATH = "profile/friend";

    public Friend(String appKey, String appSecret, RongCloud rongCloud) {
        this.appKey = appKey;
        this.appSecret = appSecret;
        this.rongCloud = rongCloud;
        initPath();
    }

    @Override
    protected void initPath() {
        super.path = API_JSON_PATH;
    }

    /**
     * Add a friend
     **/
    public ResponseResult add(FriendModel friend) throws Exception {
        String method = CheckMethod.ADD_FRIEND;

        ResponseResult result = checkFiled(friend, method, ResponseResult.class);
        if (result != null) {
            return result;
        }

        StringBuilder sb = new StringBuilder();
        addFormParam(sb, "userId=", friend.getUserId());
        addFormParam(sb, "&targetId=", friend.getTargetId());
        addFormParam(sb, "&optType=", friend.getOptType());
        addFormParam(sb, "&extra=", friend.getExtra());
        String body = sb.toString();

        return doRequest("/friend/add.json", body, method, ResponseResult.class);
    }


    /**
     * Delete a friend
     **/
    public ResponseResult delete(String userId, String... targetIds) throws Exception {
        String method = CheckMethod.DEL_FRIEND;

        ResponseResult result = checkParam("userId", userId, method, ResponseResult.class);
        if (result != null) {
            return result;
        }
        result = checkParam("targetIds", targetIds, method, ResponseResult.class);
        if (result != null) {
            return result;
        }

        StringBuilder sb = new StringBuilder();
        addFormParam(sb, "userId=", userId);
        addFormParam(sb, "&targetIds=", StringUtils.join(removeDuplicates(targetIds), ","));
        String body = sb.toString();

        return doRequest("/friend/delete.json", body, method, ResponseResult.class);
    }

    /**
     * Clear all friends
     **/
    public ResponseResult clean(String userId) throws Exception {
        String method = CheckMethod.CLEAN_FRIEND;

        ResponseResult result = checkParam("userId", userId, method, ResponseResult.class);
        if (result != null) {
            return result;
        }
        StringBuilder sb = new StringBuilder();
        addFormParam(sb, "userId=", userId);
        String body = sb.toString();

        return doRequest("/friend/clean.json", body, method, ResponseResult.class);
    }

    /**
     * Set friend profile
     **/
    public ResponseResult setProfile(FriendProfileModel profileModel) throws Exception {
        String method = CheckMethod.SET_FRIEND_PROFILE;

        ResponseResult result = checkFiled(profileModel, method, ResponseResult.class);
        if (result != null) {
            return result;
        }

        StringBuilder sb = new StringBuilder();
        addFormParam(sb, "userId=", profileModel.getUserId());
        addFormParam(sb, "&targetId=", profileModel.getTargetId());
        addFormParam(sb, "&remarkName=", profileModel.getRemarkName());
        addFormParam(sb, "&friendExtProfile=", profileModel.getFriendExtProfile());
        String body = sb.toString();

        return doRequest("/friend/profile/set.json", body, method, ResponseResult.class);
    }


    /**
     * Paginate to get friend information
     */
    public QueryFriendsResult pagingGetFriends(PagingGetFriendsModel getFriendsModel) throws Exception {
        String method = CheckMethod.PAGING_GET_FRIENDS;

        QueryFriendsResult result = checkFiled(getFriendsModel, method, QueryFriendsResult.class);
        if (result != null) {
            return result;
        }

        StringBuilder sb = new StringBuilder();
        addFormParam(sb, "userId=", getFriendsModel.getUserId());
        addFormParam(sb, "&pageToken=", getFriendsModel.getPageToken());
        addFormParam(sb, "&size=", getFriendsModel.getSize());
        addFormParam(sb, "&order=", getFriendsModel.getOrder());
        String body = sb.toString();

        return doRequest("/friend/get.json", body, method, QueryFriendsResult.class);
    }


    /**
     * Check friend relationships
     **/
    public CheckFriendsResult checkFriends(String userId, String... targetIds) throws Exception {
        String method = CheckMethod.CHECK_FRIENDS;

        CheckFriendsResult result = checkParam("userId", userId, method, CheckFriendsResult.class);
        if (result != null) {
            return result;
        }
        result = checkParam("targetIds", targetIds, method, CheckFriendsResult.class);
        if (result != null) {
            return result;
        }

        StringBuilder sb = new StringBuilder();
        addFormParam(sb, "userId=", userId);
        addFormParam(sb, "&targetIds=", StringUtils.join(removeDuplicates(targetIds), ","));
        String body = sb.toString();

        return doRequest("/friend/check.json", body, method, CheckFriendsResult.class);
    }


    /**
     * Set the permission for adding friends
     **/
    public ResponseResult setPermission(Integer permissionType, String... userIds) throws Exception {
        String method = CheckMethod.SET_PERM;

        ResponseResult result = checkParam("userIds", userIds, method, ResponseResult.class);
        if (result != null) {
            return result;
        }
        result = checkParam("permissionType", permissionType, method, ResponseResult.class);
        if (result != null) {
            return result;
        }

        StringBuilder sb = new StringBuilder();
        addFormParam(sb, "userIds=", StringUtils.join(removeDuplicates(userIds), ","));
        addFormParam(sb, "&permissionType=", permissionType);
        String body = sb.toString();

        return doRequest("/friend/permission/set.json", body, method, ResponseResult.class);
    }


    /**
     * Get the permission for adding friends
     **/
    public GetPermissionResult getPermission(String... userIds) throws Exception {
        String method = CheckMethod.GET_PERM;

        GetPermissionResult result = checkParam("userIds", userIds, method, GetPermissionResult.class);
        if (result != null) {
            return result;
        }

        StringBuilder sb = new StringBuilder();
        addFormParam(sb, "userIds=", StringUtils.join(removeDuplicates(userIds), ","));
        String body = sb.toString();

        return doRequest("/friend/permission/get.json", body, method, GetPermissionResult.class);
    }


}

File: sdk/friend.go
package sdk

import (
	"encoding/json"
	"strconv"
	"strings"
	"time"

	"github.com/astaxie/beego/httplib"
)

// FriendModel Add friend request model
type FriendModel struct {
	UserId   string `json:"userId"`   // User ID
	TargetId string `json:"targetId"` // Target user ID
	OptType  *int   `json:"optType"`  // Add type: 1: Add friend according to the target user's add friend verification level (default); 2: Add friend directly: no need to go through the target user's approval
	Extra    string `json:"extra"`    // Extra information
}

// FriendProfileModel Friend profile model
type FriendProfileModel struct {
	UserId           string `json:"userId"`           // User ID
	TargetId         string `json:"targetId"`         // Target user ID
	RemarkName       string `json:"remarkName"`       // Remark name
	FriendExtProfile string `json:"friendExtProfile"` // Friend extended profile
}

// PagingGetFriendsModel Paging get friends request model
type PagingGetFriendsModel struct {
	UserId    string `json:"userId"`    // User ID
	PageToken string `json:"pageToken"` // Pagination token
	Size      int    `json:"size"`      // Page size
	Order     *int   `json:"order"`     // Sort order (nil=not set, 0=asc, 1=desc)
}

// Response result structures for friend API
type QueryFriendsResult struct {
	TotalCount int             `json:"totalCount"` // Total number of friends
	PageToken  string          `json:"pageToken"`  // Pagination token
	Friends    []FriendProfile `json:"friendList"` // Friend list
}

type FriendProfile struct {
	UserId           string `json:"userId"`           // User ID
	Name             string `json:"name"`             // Username
	RemarkName       string `json:"remarkName"`       // Friend alias
	FriendExtProfile string `json:"friendExtProfile"` // Custom extended attributes
	Time             int64  `json:"time"`             // Timestamp when the friend was added
}

type CheckFriendsResult struct {
	Results []CheckFriendResult `json:"results"` // Friend relation list
}

type CheckFriendResult struct {
	UserId string `json:"userId"` // User ID
	Result int    `json:"result"` // Unidirectional check result: 1: Not in my friend list; 2: In my friend list; Bidirectional check result: 1: In both users' friend lists; 2: Not in both users' friend lists; 3: Only in the current user's friend list; 4: Only in the target user's friend list
}

type GetPermissionResult struct {
	PermissionSettings []PermissionSetting `json:"permissionSettings"` // User permission list
}

type PermissionSetting struct {
	UserId string `json:"userId"` // User ID
	Type   int    `json:"type"`   // Permission type
}

// Helper function: validate user ID for friend operations
func (rc *RongCloud) validateFriendUserId(userId string) error {
	if userId == "" {
		return RCErrorNew(1002, "Parameter 'userId' is required")
	}
	return nil
}

// Helper function: validate target IDs
func (rc *RongCloud) validateTargetIds(targetIds []string) error {
	if len(targetIds) == 0 {
		return RCErrorNew(1002, "Parameter 'targetIds' is required")
	}
	return nil
}

// Add friend
func (rc *RongCloud) FriendAdd(friend FriendModel) (ResponseResult, error) {
	result := ResponseResult{}

	// Validate required parameters
	if err := rc.validateFriendUserId(friend.UserId); err != nil {
		return result, err
	}
	if friend.TargetId == "" {
		return result, RCErrorNew(1002, "Parameter 'targetId' is required")
	}

	req := httplib.Post(rc.rongCloudURI + "/friend/add.json")
	req.SetTimeout(time.Second*rc.timeout, time.Second*rc.timeout)
	rc.fillHeader(req)

	req.Param("userId", friend.UserId)
	req.Param("targetId", friend.TargetId)
	if friend.OptType != nil {
		req.Param("optType", strconv.Itoa(*friend.OptType))
	}
	if friend.Extra != "" {
		req.Param("extra", friend.Extra)
	}

	resp, err := rc.do(req)
	if err != nil {
		rc.urlError(err)
		return result, err
	}

	if err := json.Unmarshal(resp, &result); err != nil {
		return result, err
	}

	return result, nil
}

// Delete friend
func (rc *RongCloud) FriendDelete(userId string, targetIds ...string) (ResponseResult, error) {
	result := ResponseResult{}

	// Validate required parameters
	if err := rc.validateFriendUserId(userId); err != nil {
		return result, err
	}
	if err := rc.validateTargetIds(targetIds); err != nil {
		return result, err
	}

	req := httplib.Post(rc.rongCloudURI + "/friend/delete.json")
	req.SetTimeout(time.Second*rc.timeout, time.Second*rc.timeout)
	rc.fillHeader(req)

	req.Param("userId", userId)
	req.Param("targetIds", strings.Join(removeDuplicates(targetIds), ","))

	resp, err := rc.do(req)
	if err != nil {
		rc.urlError(err)
		return result, err
	}

	if err := json.Unmarshal(resp, &result); err != nil {
		return result, err
	}

	return result, nil
}

// Clean all friends
func (rc *RongCloud) FriendClean(userId string) (ResponseResult, error) {
	result := ResponseResult{}

	// Validate required parameters
	if err := rc.validateFriendUserId(userId); err != nil {
		return result, err
	}

	req := httplib.Post(rc.rongCloudURI + "/friend/clean.json")
	req.SetTimeout(time.Second*rc.timeout, time.Second*rc.timeout)
	rc.fillHeader(req)

	req.Param("userId", userId)

	resp, err := rc.do(req)
	if err != nil {
		rc.urlError(err)
		return result, err
	}

	if err := json.Unmarshal(resp, &result); err != nil {
		return result, err
	}

	return result, nil
}

// Set friend profile
func (rc *RongCloud) FriendSetProfile(profileModel FriendProfileModel) (ResponseResult, error) {
	result := ResponseResult{}

	// Validate required parameters
	if err := rc.validateFriendUserId(profileModel.UserId); err != nil {
		return result, err
	}
	if profileModel.TargetId == "" {
		return result, RCErrorNew(1002, "Parameter 'targetId' is required")
	}

	req := httplib.Post(rc.rongCloudURI + "/friend/profile/set.json")
	req.SetTimeout(time.Second*rc.timeout, time.Second*rc.timeout)
	rc.fillHeader(req)

	req.Param("userId", profileModel.UserId)
	req.Param("targetId", profileModel.TargetId)
	if profileModel.RemarkName != "" {
		req.Param("remarkName", profileModel.RemarkName)
	}
	if profileModel.FriendExtProfile != "" {
		req.Param("friendExtProfile", profileModel.FriendExtProfile)
	}

	resp, err := rc.do(req)
	if err != nil {
		rc.urlError(err)
		return result, err
	}

	if err := json.Unmarshal(resp, &result); err != nil {
		return result, err
	}

	return result, nil
}

// Paginate to get friend information
func (rc *RongCloud) PagingGetFriends(getFriendsModel PagingGetFriendsModel) (QueryFriendsResult, error) {
	result := QueryFriendsResult{}

	// Validate required parameters
	if err := rc.validateFriendUserId(getFriendsModel.UserId); err != nil {
		return result, err
	}

	req := httplib.Post(rc.rongCloudURI + "/friend/get.json")
	req.SetTimeout(time.Second*rc.timeout, time.Second*rc.timeout)
	rc.fillHeader(req)

	req.Param("userId", getFriendsModel.UserId)
	if getFriendsModel.PageToken != "" {
		req.Param("pageToken", getFriendsModel.PageToken)
	}
	if getFriendsModel.Size > 0 {
		req.Param("size", strconv.Itoa(getFriendsModel.Size))
	}
	if getFriendsModel.Order != nil {
		req.Param("order", strconv.Itoa(*getFriendsModel.Order))
	}

	resp, err := rc.do(req)
	if err != nil {
		rc.urlError(err)
		return result, err
	}

	if err := json.Unmarshal(resp, &result); err != nil {
		return result, err
	}

	return result, nil
}

// Check friend relationships
func (rc *RongCloud) FriendCheckFriends(userId string, targetIds ...string) (CheckFriendsResult, error) {
	result := CheckFriendsResult{}

	// Validate required parameters
	if err := rc.validateFriendUserId(userId); err != nil {
		return result, err
	}
	if err := rc.validateTargetIds(targetIds); err != nil {
		return result, err
	}

	req := httplib.Post(rc.rongCloudURI + "/friend/check.json")
	req.SetTimeout(time.Second*rc.timeout, time.Second*rc.timeout)
	rc.fillHeader(req)

	req.Param("userId", userId)
	req.Param("targetIds", strings.Join(removeDuplicates(targetIds), ","))

	resp, err := rc.do(req)
	if err != nil {
		rc.urlError(err)
		return result, err
	}

	if err := json.Unmarshal(resp, &result); err != nil {
		return result, err
	}

	return result, nil
}

// Set permission for adding friends
func (rc *RongCloud) FriendSetPermission(permissionType int, userIds ...string) (ResponseResult, error) {
	result := ResponseResult{}

	// Validate required parameters
	if len(userIds) == 0 {
		return result, RCErrorNew(1002, "Parameter 'userIds' is required")
	}
	if permissionType == 0 {
		return result, RCErrorNew(1002, "Parameter 'permissionType' is required")
	}

	req := httplib.Post(rc.rongCloudURI + "/friend/permission/set.json")
	req.SetTimeout(time.Second*rc.timeout, time.Second*rc.timeout)
	rc.fillHeader(req)

	req.Param("userIds", strings.Join(removeDuplicates(userIds), ","))
	req.Param("permissionType", strconv.Itoa(permissionType))

	resp, err := rc.do(req)
	if err != nil {
		rc.urlError(err)
		return result, err
	}

	if err := json.Unmarshal(resp, &result); err != nil {
		return result, err
	}

	return result, nil
}

// Get permission for adding friends
func (rc *RongCloud) FriendGetPermission(userIds ...string) (GetPermissionResult, error) {
	result := GetPermissionResult{}

	// Validate required parameters
	if len(userIds) == 0 {
		return result, RCErrorNew(1002, "Parameter 'userIds' is required")
	}

	req := httplib.Post(rc.rongCloudURI + "/friend/permission/get.json")
	req.SetTimeout(time.Second*rc.timeout, time.Second*rc.timeout)
	rc.fillHeader(req)

	req.Param("userIds", strings.Join(removeDuplicates(userIds), ","))

	resp, err := rc.do(req)
	if err != nil {
		rc.urlError(err)
		return result, err
	}

	if err := json.Unmarshal(resp, &result); err != nil {
		return result, err
	}

	return result, nil
}
